### ISSUE TITLE
Training on custom CARLA dataset

### ISSUE THREAD

**Issue by flomllr** [nianticlabs/monodepth2#130]
Hi there,
great project!
I'm trying to retrain with a monocular dataset generated using the CARLA simulator. This is how the training data looks like:
![rgb](https://user-images.githubusercontent.com/29047312/72217868-3ca15900-3534-11ea-9dd4-b79b0548c418.gif)

For this purpose, I implemented a new Dataset inheriting from MonoDataset:
```python
import numpy as np
import PIL.Image as pil
import os

from .mono_dataset import MonoDataset

# FOV
FOV = 90

# image size
img_width = 200
img_height = 100

image_size = [img_width, img_height]

class CARLADataset(MonoDataset):
    def __init__(self, *args, **kwargs):
        super(CARLADataset, self).__init__(*args, **kwargs)

        self.K = np.array([[img_width / (2 * np.tan(FOV * np.pi / 360)), 0, img_width / 2, 0],
                          [0, img_width / (2 * np.tan(FOV * np.pi / 360)), img_height / 2, 0],
                          [0, 0, 1, 0],
                          [0, 0, 0, 1]], dtype=np.float32)

        self.full_res_shape = (img_width, img_height)

    def check_depth(self):
        return False

    def get_image_path(self, folder, frame_index, side):
        f_str = "image_{:05d}.png".format(frame_index)
        image_path = os.path.join(self.data_path, folder, f_str)
        return image_path

    def get_color(self, folder, frame_index, side, do_flip):
        color = self.loader(self.get_image_path(folder, frame_index, side))

        if do_flip:
            color = color.transpose(pil.FLIP_LEFT_RIGHT)
        
        return color
```

I then trained a model for 250 epochs on a dataset containing 1300 frames - this is a resulting prediction:
![model1](https://user-images.githubusercontent.com/29047312/72217915-e2ed5e80-3534-11ea-9e23-606c50c8ef0b.jpeg)

I also tried training for 5000 episodes on a smaller dataset only containing 135 frames - here is the result:
![model2](https://user-images.githubusercontent.com/29047312/72217939-2ba51780-3535-11ea-83b5-fafefb25867f.jpeg)

Here are the options I used for the training:
```bash
python monodepth2/train.py \
    --model_name carla_mono \
    --data_path=data \
    --split=carla \
    --dataset=carla \
    --png \
    --height=96 \
    --width=192 \
    --max_depth=100 \
    --num_epochs 20000 \
    --log_frequency=50
```
(For the first try I used a max_depth of 1000, for the second try a max_depth of 100.)

Edit: I just tried finetuning the mono_640x192 model.
Before finetuning:
![before](https://user-images.githubusercontent.com/29047312/72218531-ec2df980-353b-11ea-8a12-2e040e771afc.jpeg)

After 2 episodes of finetuning:
![image_00001_disp](https://user-images.githubusercontent.com/29047312/72218532-efc18080-353b-11ea-99ad-632a81ad1bc9.jpeg)

I'm grateful for any hints towards what I might be doing wrong!

**Comment by mrharicot** [comment]
Hi,

Here are the issues I can think about:
- This is very likely it. Your intrinsics matrix is incorrect it should be normalized by the image size.
```python
focal = img_width / (2 * np.tan(FOV * np.pi / 360))
self.K = np.array([[focal / img_width, 0, 0.5, 0],
                  [0, focal / img_height, 0.5, 0],
                  [0, 0, 1, 0],
                  [0, 0, 0, 1]], dtype=np.float32)
```
- 1300 frames is very small, you likely need more data.
- You image resolution is very small. While it isn't necessarily a problem, it becomes one if you render the images at such a low res because of aliasing. As you can see on the gif you shared, there is a lot of shimmering visible resulting in "fake" motion, which is uncorrelated to actual rigid motion. This is also an issue as we use a reprojection loss, and even if we had ground truth depth the reprojection would not be zero (ignoring occlusions) because of aliasing. I recommend rendering them at a higher resolution then downsampling them appropriately.

**Comment by flomllr** [comment]
Thank you so much! Your hints fixed the issues.

I sampled 9500 anti-aliased images in 640x192 resolution. Here is the result after training for 19 epochs from scratch:
![test_simple](https://user-images.githubusercontent.com/29047312/72354559-48cd1800-36e6-11ea-9a27-a3fa0f35c08c.gif)

**Comment by caikw0602** [comment]
Hi. I wanna train my own monocular dataset with monodepth2 too. But I am struggling how to make the right data orgarization. Further, I do not know how to name the images of different frames. Would you please give some help? Some work steps? Thank you very much.

**Comment by flomllr** [comment]
Sure! I organized my dataset in the following way:
```
data_path/

    episode_000/
        HD_RGB2/
            image_00000.png
            image_00001.png
            ...

    episode_001/
        HD_RGB2/
            image_00000.png
            ...
    ...
```

And then the train_files.txt file looks like this:

```
episode_000/HD_RGB2 1 l
episode_000/HD_RGB2 2 l
episode_000/HD_RGB2 3 l
...
episode_001/HD_RGB2 1 l
...
```

Since it's a monocular dataset, all images are marked with `l`

Depending on your structure you have to adjust the `get_image_path` method in your dataset class. Mine looks like this:
```python
def get_image_path(self, folder, frame_index, side):
    f_str = "image_{:05d}.png".format(frame_index)
    image_path = os.path.join(self.data_path, folder, f_str)
    return image_path
```

Hope this helps.

**Comment by Jinho-P** [comment]
Hi, @flomllr 

could I ask some question?

1. What is resolution of your image taken from carla simulator. Is it same as kitti image [1242, 375]?
2. Do you resize original image from carla to specific size and specific ratio? 
3. Your last result looks like really great. For this results, are there any other methods? (except change intrinsic matrix to normalized intrinsic matrix)
